### PR TITLE
Trace element

### PR DESCRIPTION
**Pull request by thomasgibson**
This branch depends on the following other branches: firedrakeproject/tsfc#46, firedrakeproject/ufl#3, and firedrakeproject/fiat#6.

Summary of changes:
1. `tests/regression/test_hybridisation.py` is now using the new `TraceHDiv` element from FIAT.
2. `tests/regression/test_trace_galerkin_projection.py` is test for making sure the trace space is producing expected results and convergence rates.

### CODE REVIEW

**Comment by miklos1** [comment]
Now that you have filed all pull request, you can edit the descriptions to reference the pull requests in the other projects instead of just naming the branches.

File: firedrake/functionspacedata.py [hunk 1]
```diff
@@ -552,7 +552,7 @@ def get_shared_data(mesh, fiat_element):
     """
     if not isinstance(mesh, mesh_mod.MeshTopology):
         raise ValueError("%s is not a MeshTopology" % mesh)
-    if not isinstance(fiat_element, FIAT.finite_element.FiniteElement):
+    if not isinstance(fiat_element, (FIAT.finite_element.CiarletElement, FIAT.finite_element.FiniteElement)):
```
Comment (miklos1) [inline firedrake/functionspacedata.py]: Doesn't `CiarletElement` inherit from `FiniteElement`? It should. If it does, it should sufficient to check against FiniteElement only.

File: firedrake/tsfc_interface.py [hunk 2]
```diff
@@ -103,6 +106,17 @@ def split(self, form):
     def multi_index(self, o):
         return o
 
+    def indexed(self, o, expr, multiindex):
```
Comment (miklos1) [inline firedrake/tsfc_interface.py]: Is this still useful?

Comment (thomasgibson) [inline firedrake/tsfc_interface.py]: I just ran some tests without this bit. I think it's no longer useful. I'll remove it in my next commit.

**Comment by miklos1** [comment]
Too many commits for a one-line change. Maybe you could squash them.

**Comment by miklos1** [comment]
Have you considered writing more tests for the trace class?

**Comment by thomasgibson** [comment]
Yeah I have them. I'll be uploading some tests shortly

**Comment by thomasgibson** [comment]
Squashed all old and irrelevant commits.

**Comment by thomasgibson** [comment]
Y'all seemed to disapprove when I suggested relaxing by a factor of 10, so I opted for minimal relaxation. But perhaps I just misinterpreted. I don't mind changing it to `1e-13` if that's preferable.

**Pull request merged**